### PR TITLE
docs: Add best practices documentation

### PR DESCRIPTION
**Pull request by damassi**
Adds some documentation around Force best practices. (h/t to Eigen's docs for providing some of the structure here.) 

I tried to keep things to point and current, but I am very likely missing stuff -- thoughts? 

Note that this is something of a launch point for adding more detail in our currently-stubbed-out [playbook](https://www.notion.so/artsy/Frontend-f7946aca868842ad8e758809826b558e). 

Tagged a few folks with extensive force experience for feedback, but if you feel someone else might have more ideas please tag them as well 🙏

### CODE REVIEW

File: docs/best_practices.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+- [Artsy JavaScriptures seminar on Relay](https://github.com/artsy/javascriptures/tree/master/4_intro-to-relay)
+- Examples
+  - [A top-level route-based Relay request](https://github.com/artsy/force/blob/0b291f005763e7c2600a5077786c9510bf655079/src/v2/Apps/Consign/consignRoutes.tsx#L28-L34)
+  - [A fragment container](https://github.com/artsy/force/blob/0b291f005763e7c2600a5077786c9510bf655079/src/v2/Apps/Consign/Routes/Offer/OfferDetailApp.tsx#L47-L57)
+
+### Prefer Relay containers (higher order components) over hooks
+
+We have a preference for Relay containers due to [`relay-hooks`](https://github.com/relay-tools/relay-hooks) hooks not being compatible with Relay containers which represent the majority of our components using Relay. (This could change once Relay releases its official hooks implementation.)
+
+### Keep file structure organized
```
Comment (dzucconi) [inline docs/best_practices.md:46]: Something about not renaming named imports might be a good addition too.

**Review by dzucconi — APPROVED** [review]

File: docs/best_practices.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+### When writing UI, use Palette
+
+Our design system [Palette](https://github.com/artsy/palette) is used for most UI needs. Check out [the docs](https://palette.artsy.net/)!
+
+> If interested in some of the lower-level particulars of how Palette was built, see the [styled-system](https://styled-system.com/) docs.
+
+### For routing, use our framework
+
+Individual sub-apps (represented by routes like `/artist/:id` or `/collect`) are built on top of a home-grown SSR (server-side-rendering) framework based around [Found](https://github.com/4Catalyzer/found), a routing library. Sub-apps are mounted within the [global routes file](https://github.com/artsy/force/blob/1842553ad34475bc3b804f00c6410d7f23d64f65/src/v2/Apps/getAppRoutes.tsx).
```
Comment (ashfurrow) [inline docs/best_practices.md:24]: > home-grown SSR

This is Fresnel, right? Should we link to it here?

File: docs/best_practices.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+Data should be loaded from [Metaphysics](https://github.com/artsy/metaphysics), Artsy's GraphQL server. Requests to Metaphysics should be made through [Relay](https://relay.dev).
+
+- [Why Artsy uses Relay](http://artsy.github.io/blog/2017/02/05/Front-end-JavaScript-at-Artsy-2017/#Relay)
+- [Artsy JavaScriptures seminar on Relay](https://github.com/artsy/javascriptures/tree/master/4_intro-to-relay)
+- Examples
+  - [A top-level route-based Relay request](https://github.com/artsy/force/blob/0b291f005763e7c2600a5077786c9510bf655079/src/v2/Apps/Consign/consignRoutes.tsx#L28-L34)
+  - [A fragment container](https://github.com/artsy/force/blob/0b291f005763e7c2600a5077786c9510bf655079/src/v2/Apps/Consign/Routes/Offer/OfferDetailApp.tsx#L47-L57)
+
+### Prefer Relay containers (higher order components) over hooks
```
Comment (ashfurrow) [inline docs/best_practices.md:42]: My understanding is that hooks are still used in functional components, just not in Relay fragment containers. Is that right?

**Review by ashfurrow — COMMENTED** [review]
Looks great! Just a question, and a suggestion 🙏

**Review by jonallured — APPROVED** [review]
Awesome work, thanks for starting this!! ❤️

File: docs/best_practices.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+Framework code is located in [`v2/Artsy`](https://github.com/artsy/force/tree/1842553ad34475bc3b804f00c6410d7f23d64f65/src/v2/Artsy).
+
+### Write unit tests for new components
+
+We use [Jest](https://jestjs.io/) for our unit tests, and [Mocha](https://mochajs.org/) for testing legacy code.
+
+Some top-level notes:
+
+- We use [`enzyme`](https://enzymejs.github.io/enzyme/)
+- We don't like snapshot tests; they produce too much churn for too little value.
```
Comment (pepopowitz) [inline docs/best_practices.md:88]: Non-blocking: use a verb that is less subjective. I also don't like snapshot tests, but I don't think it is a universal opinion. There is no question that we *avoid* them though. Removing the subjectivity of this sentence should make it less likely for people to want to rewrite it.

```suggestion
- We avoid snapshot tests; they produce too much churn for too little value.
```

File: docs/best_practices.md (hunk 1 above)
Comment (damassi) [reply docs/best_practices.md:46]: Can you elaborate a bit? (Or do you mean not using default exports?)

File: docs/best_practices.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+### Keep file structure organized
+
+Generally speaking, all new product-centric code is written [inside the `src/v2`](https://github.com/artsy/force/tree/1842553ad34475bc3b804f00c6410d7f23d64f65/src/v2) folder.
+
+Sub-apps are written [inside of `src/v2/Apps` folder](https://github.com/artsy/force/tree/1842553ad34475bc3b804f00c6410d7f23d64f65/src/v2/Apps).
+
+Within a sub-app, things are typically structured like so:
+
+```
+├── routes.tsx
```
Comment (pepopowitz) [inline docs/best_practices.md:55]: 💪 I'm really happy to see this documented, to remove individual discretion from code organization.

File: docs/best_practices.md (hunk 2 above)
Comment (damassi) [reply docs/best_practices.md:24]: Not fresnel, fresnel is more of a UI tool for rendering responsive layouts on the server - this is more about our app framework / routing / bundle splitting, etc

**Review by pepopowitz — APPROVED** [review]
Amazing!!! Thank you!

I added one suggestion to remove subjectivity but I certainly wouldn't hold up a merge for it.

File: docs/best_practices.md (hunk 1 above)
Comment (dzucconi) [reply docs/best_practices.md:46]: I think there's a lint rule for it now, so I guess mentioning it isn't necessary. Was just talking about like preferring `import { FooFragmentContainer } from 'Foo'` over `import { FooFragmentContainer as Foo } from 'Foo'`

File: docs/best_practices.md (hunk 3 above)
Comment (damassi) [reply docs/best_practices.md:42]: We use the HOC pattern everywhere in Force (I believe this is also the case in Eigen, too). I think clarifying this as `over relay-hooks` in the header will help

File: docs/best_practices.md (hunk 1 above)
Comment (damassi) [reply docs/best_practices.md:46]: Ah, totally, yup that makes sense. I'll add, lint rule or no, as this is now something of an artsy-wide best practice. Docs will reenforce 👍

File: docs/best_practices.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+- Example code.
+- Pull requests with interesting discussions.
+- Conversations on Slack.
+- Blog posts.
+
+Links should point to specific commits, and not a branch (in case the branch or file is deleted, these links should always work). But it's possible that a file is outdated, that our understanding has moved on since it was linked to; in that case, please update this document.
+
+## Current Preferred Practices
+
+The app is currently written responsively in React, with an ever-increasing amount of legacy code written in Backbone and Coffeescript. As of Nov 2020, most of our main pages have been migrated, and those [that remain become fewer](https://github.com/artsy/force/pull/6682) with time. Server-side code is built on top of Express.js; however, most server-side needs have been abstracted away by our framework (see below).
```
Comment (anandaroop) [inline docs/best_practices.md:14]: This should be "ever-*de*creasing amount of legacy code", right? 😅 

```suggestion
The app is currently written responsively in React, with an ever-decreasing amount of legacy code written in Backbone and Coffeescript. As of Nov 2020, most of our main pages have been migrated, and those [that remain become fewer](https://github.com/artsy/force/pull/6682) with time. Server-side code is built on top of Express.js; however, most server-side needs have been abstracted away by our framework (see below).
```

File: docs/best_practices.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+Our design system [Palette](https://github.com/artsy/palette) is used for most UI needs. Check out [the docs](https://palette.artsy.net/)!
+
+> If interested in some of the lower-level particulars of how Palette was built, see the [styled-system](https://styled-system.com/) docs.
+
+### For routing, use our framework
+
+Individual sub-apps (represented by routes like `/artist/:id` or `/collect`) are built on top of a home-grown SSR (server-side-rendering) framework based around [Found](https://github.com/4Catalyzer/found), a routing library. Sub-apps are mounted within the [global routes file](https://github.com/artsy/force/blob/1842553ad34475bc3b804f00c6410d7f23d64f65/src/v2/Apps/getAppRoutes.tsx).
+
+To learn how to create a new sub-app, see [the docs](https://github.com/artsy/force/blob/1842553ad34475bc3b804f00c6410d7f23d64f65/docs/adding_new_app.md).
```
Comment (anandaroop) [inline docs/best_practices.md:26]: Is this intentionally a permalink to an old version of this document, or should it link to the evergreen version?

```suggestion
To learn how to create a new sub-app, see [the docs](https://github.com/artsy/force/blob/master/docs/adding_new_app.md).
```

File: docs/best_practices.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+We use [Jest](https://jestjs.io/) for our unit tests, and [Mocha](https://mochajs.org/) for testing legacy code.
+
+Some top-level notes:
+
+- We use [`enzyme`](https://enzymejs.github.io/enzyme/)
+- We avoid snapshot tests; they produce too much churn for too little value.
+- We use the `relay-test-utils` package for testing Relay code, and [this helper](https://github.com/artsy/force/blob/0b291f005763e7c2600a5077786c9510bf655079/src/v2/DevTools/setupTestWrapper.tsx) for quickly spinning up tests.
+
+Here are some great examples of what tests and test coverage should look like.
```
Comment (anandaroop) [inline docs/best_practices.md:91]: 🙏 💯 this is thing I'm always craving.

**Review by anandaroop — APPROVED** [review]
Thank you!

File: docs/best_practices.md (hunk 7 above)
Comment (damassi) [reply docs/best_practices.md:26]: It was intentional to link to something concrete. Having a link that might be slightly outdated (but representative within the context of the surrounding documentation) is preferrable to a 404 (which happens frequently). I think as this evolves these links will evolve too (when necessary).

File: docs/best_practices.md (hunk 4 above)
Comment (damassi) [reply docs/best_practices.md:126]: ugh @pepopowitz - i thought i hit commit suggestion here, updating

Comment (damassi) [reply docs/best_practices.md:126]: Updated here: https://github.com/artsy/force/commit/77c6c5c3512a981e2ca183a5ba2a2b2e4a5e2987

**Pull request merged**